**EXERCISE-3**

In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.1 MB 8.0 MB/s 
     |████████████████████████████████| 49 kB 5.6 MB/s 


In [2]:
from pyomo.environ import *

In [3]:
import numpy as np
import pandas as pd

In [4]:
data_csv=pd.read_csv('lab4_ex3.csv')

#Que.3(4)

Let $x_i's$ be the number of drones of model i, where $i\in \{1,2,3,4,5\}$ and 

let $x_1$ is the number of drones of MANAS 

 $x_2$ is the number of drones of JAY
 
$x_3$ is the number of drones of VEER 

$x_4$ is the number of drones of SPASHT 

and $x_5$ is the number of drones of DHAKSH.


Now, as we want to maximize total carrying capacity. Hence, the Optimization model will be:
$$ \text{Max} \  12x_1+14x_2+10x_3+15x_4+14x_5\\
\text{s.t.} \ \ \ \  \ \ \  6x_1+7.5x_2+9.5x_3+10x_4+12.5x_5 \leq 59 \\
\ \ \ \ \ \ \ 4.25x_1+3.25x_2+2.5x_3+3.45x_4+2.25x_5 \leq 73 \\ x_i \in \mathbb{Z^+} \text{for} \ i\in\{1,2,...,5\} $$

In this problem $x_i's$ must be non negative integers as the number of drones can't be negative or in decimal.

In [5]:
data_csv.columns

Index(['Drone Model', 'Carrying Capacity', 'Cost', 'Parking Space'], dtype='object')

In [6]:
model=ConcreteModel()

In [7]:
N=len(data_csv.index)#N denotes the number of decision variables

In [8]:
M=2#M denotes the number of constraints

In [9]:
col_indices=range(N)
row_indices=range(M)

In [10]:
model.x=Var(col_indices,domain=NonNegativeIntegers)

In [11]:
model.objective=Objective(expr=summation(data_csv['Carrying Capacity'],model.x),sense=maximize)

In [12]:
model.constraints=ConstraintList()

In [13]:
model.constraints.add(sum(data_csv['Cost'][j]*model.x[j] for j in col_indices)<=59)
model.constraints.add(sum(data_csv['Parking Space'][j]*model.x[j] for j in col_indices)<=73)

In [14]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True : NonNegativeIntegers
          1 :     0 :  None :  None : False :  True : NonNegativeIntegers
          2 :     0 :  None :  None : False :  True : NonNegativeIntegers
          3 :     0 :  None :  None : False :  True : NonNegativeIntegers
          4 :     0 :  None :  None : False :  True : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 12*x[0] + 14*x[1] + 10*x[2] + 15*x[3] + 

In [15]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 148489 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [16]:
opt_cbc=SolverFactory('cbc')

In [17]:
result1=opt_cbc.solve(model)



In [18]:
print('Solver status:', result1.solver.status)
print('Solver termination condition:',result1.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


#Que.3(6)

**optimal value and value of the decision variables:-**

In [19]:
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  114.0

Decision Variables
MANAS  =  6.0
JAY  =  3.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=2
    Key : Lower : Body  : Upper
      1 :  None :  58.5 :  59.0
      2 :  None : 35.25 :  73.0


#Que.3(7)

**removing the restrictions of decision variables:**

In [20]:
model.x.domain=NonNegativeReals

In [21]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   6.0 :  None : False : False : NonNegativeReals
          1 :     0 :   3.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 12*x[0] + 14*x[1] + 10*x[2] + 15*x[3] + 14*x[4]

1 Cons

In [22]:
result2=opt_cbc.solve(model)
print('Solver status:', result2.solver.status)
print('Solver termination condition:',result2.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


**optimal value and the value of the decision variables:**

In [23]:
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  117.9999996

Decision Variables
MANAS  =  9.8333333
JAY  =  0.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=2
    Key : Lower : Body         : Upper
      1 :  None :   58.9999998 :  59.0
      2 :  None : 41.791666525 :  73.0


#Que.3(8)

No the solution of the MILP can't be obtained by merely rounding the solution of the LP.because here we are getting the number of MANAS drone is 9.83333 if we round off this then it will be 10 then the cost of buying 10 MANAS drone will go to 60 crore which doesn't satisfy the budget constraint of 59 crore.



#Que.3(9)

**solving new LP after increasing the RHS side in our constraints:-**

In [24]:
for i in range(2):
  model.constraints[i+1].deactivate()

In [25]:
model.constraints.add(sum(data_csv['Cost'][j]*model.x[j] for j in col_indices)<=71)
model.constraints.add(sum(data_csv['Parking Space'][j]*model.x[j] for j in col_indices)<=87)

In [26]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 9.8333333 :  None : False : False : NonNegativeReals
          1 :     0 :       0.0 :  None : False : False : NonNegativeReals
          2 :     0 :       0.0 :  None : False : False : NonNegativeReals
          3 :     0 :       0.0 :  None : False : False : NonNegativeReals
          4 :     0 :       0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 12*x[0] + 14*x[1] + 10*x[

In [27]:
result3=opt_cbc.solve(model)
print('Solver status:', result3.solver.status)
print('Solver termination condition:',result3.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


**optimal value and values of the decision variables of new LP:**

In [28]:
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  141.999996

Decision Variables
MANAS  =  11.833333
JAY  =  0.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=4
    Key : Lower : Body        : Upper
      3 :  None :   70.999998 :  71.0
      4 :  None : 50.29166525 :  87.0


After increasing the RHS side of the constraints the optimal value of the new LP has increases and also the number of the MANAS drone increases.

**Solving new MILP After increasing the RHS of the constraints:**

In [29]:
model.x.domain=NonNegativeIntegers

In [30]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.833333 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 12*x[0] + 

In [31]:
result4=opt_cbc.solve(model)
print('Solver status:', result3.solver.status)
print('Solver termination condition:',result3.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


**optimal value and values of the decision variables of new MILP:**

In [32]:
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  138.0

Decision Variables
MANAS  =  8.0
JAY  =  3.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=4
    Key : Lower : Body  : Upper
      3 :  None :  70.5 :  71.0
      4 :  None : 43.75 :  87.0


The optimal objective value has increased and no. of  MANAS drones has increased.

#Que.3(10)


**solving for the given first pair that is 67(budget) and 88(space)**

In [33]:
model.constraints[3].deactivate()
model.constraints[4].deactivate()

In [34]:
model.constraints.add(sum(data_csv['Cost'][j]*model.x[j] for j in col_indices)<=67)
model.constraints.add(sum(data_csv['Parking Space'][j]*model.x[j] for j in col_indices)<=88)

In [35]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   8.0 :  None : False : False : NonNegativeIntegers
          1 :     0 :   3.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :   0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 12*x[0] + 14*x[1] + 10*x[2] 

#Solving MILP:-

In [36]:
result5=opt_cbc.solve(model)
print('Solver status:', result3.solver.status)
print('Solver termination condition:',result3.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [37]:
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  132.0

Decision Variables
MANAS  =  11.0
JAY  =  0.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=6
    Key : Lower : Body  : Upper
      5 :  None :  66.0 :  67.0
      6 :  None : 46.75 :  88.0


#Solving LP:-

In [38]:
model.x.domain=NonNegativeReals

In [39]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {1, 2, 3, 4, 5, 6}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  11.0 :  None : False : False : NonNegativeReals
          1 :     0 :   0.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 12*x[0] + 14*x[1] + 10*x[2] + 15*x[3] + 14*

In [40]:
result6=opt_cbc.solve(model)
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  134.000004

Decision Variables
MANAS  =  11.166667
JAY  =  0.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=6
    Key : Lower : Body        : Upper
      5 :  None :   67.000002 :  67.0
      6 :  None : 47.45833475 :  88.0


**solving for the second pair that is 65(budget) and 89(space)**

#Solving LP:-

In [41]:
model.constraints[5].deactivate()
model.constraints[6].deactivate()
model.constraints.add(sum(data_csv['Cost'][j]*model.x[j] for j in col_indices)<=65)
model.constraints.add(sum(data_csv['Parking Space'][j]*model.x[j] for j in col_indices)<=89)
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    8 : {1, 2, 3, 4, 5, 6, 7, 8}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 11.166667 :  None : False : False : NonNegativeReals
          1 :     0 :       0.0 :  None : False : False : NonNegativeReals
          2 :     0 :       0.0 :  None : False : False : NonNegativeReals
          3 :     0 :       0.0 :  None : False : False : NonNegativeReals
          4 :     0 :       0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 12*x[0] + 14*

In [42]:
result7=opt_cbc.solve(model)
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  129.999996

Decision Variables
MANAS  =  10.833333
JAY  =  0.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=8
    Key : Lower : Body        : Upper
      7 :  None :   64.999998 :  65.0
      8 :  None : 46.04166525 :  89.0


#Solving MILP:-

In [43]:
model.x.domain=NonNegativeIntegers
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    8 : {1, 2, 3, 4, 5, 6, 7, 8}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 10.833333 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize 

In [44]:
result8=opt_cbc.solve(model)
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  126.0

Decision Variables
MANAS  =  7.0
JAY  =  3.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=8
    Key : Lower : Body : Upper
      7 :  None : 64.5 :  65.0
      8 :  None : 39.5 :  89.0


**Solving for the third pair that is 63(budget) and 90(space)**

#Solving MILP:-

In [45]:
model.constraints[7].deactivate()
model.constraints[8].deactivate()
model.constraints.add(sum(data_csv['Cost'][j]*model.x[j] for j in col_indices)<=63)
model.constraints.add(sum(data_csv['Parking Space'][j]*model.x[j] for j in col_indices)<=90)
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   7.0 :  None : False : False : NonNegativeIntegers
          1 :     0 :   3.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :   0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :   0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 12*x[0] + 14*x[

In [46]:
result9=opt_cbc.solve(model)
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  124.0

Decision Variables
MANAS  =  8.0
JAY  =  2.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=10
    Key : Lower : Body : Upper
      9 :  None : 63.0 :  63.0
     10 :  None : 40.5 :  90.0


#Solving LP:-

In [47]:
model.x.domain=NonNegativeReals

In [48]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   10 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :   8.0 :  None : False : False : NonNegativeReals
          1 :     0 :   2.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 12*x[0] + 14*x[1] + 10*x[2] + 

In [49]:
result10=opt_cbc.solve(model)
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  126.0

Decision Variables
MANAS  =  10.5
JAY  =  0.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=10
    Key : Lower : Body   : Upper
      9 :  None :   63.0 :  63.0
     10 :  None : 44.625 :  90.0


**solving for the 4th pair that is 61(budget) and 91(space):-**

#Solving LP:-

In [50]:
model.constraints[9].deactivate()
model.constraints[10].deactivate()
model.constraints.add(sum(data_csv['Cost'][j]*model.x[j] for j in col_indices)<=61)
model.constraints.add(sum(data_csv['Parking Space'][j]*model.x[j] for j in col_indices)<=91)
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  10.5 :  None : False : False : NonNegativeReals
          1 :     0 :   0.0 :  None : False : False : NonNegativeReals
          2 :     0 :   0.0 :  None : False : False : NonNegativeReals
          3 :     0 :   0.0 :  None : False : False : NonNegativeReals
          4 :     0 :   0.0 :  None : False : False : NonNegativeReals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 12*x[0] + 14*x[1] + 10

In [51]:
result11=opt_cbc.solve(model)
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  122.000004

Decision Variables
MANAS  =  10.166667
JAY  =  0.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=12
    Key : Lower : Body        : Upper
     11 :  None :   61.000002 :  61.0
     12 :  None : 43.20833475 :  91.0


#solving MILP:-

In [52]:
model.x.domain=NonNegativeIntegers
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {0, 1, 2, 3, 4}

1 Var Declarations
    x : Size=5, Index=x_index
        Key : Lower : Value     : Upper : Fixed : Stale : Domain
          0 :     0 : 10.166667 :  None : False : False : NonNegativeIntegers
          1 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          2 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          3 :     0 :       0.0 :  None : False : False : NonNegativeIntegers
          4 :     0 :       0.0 :  None : False : False : NonNegativeIntegers

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   T

In [53]:
result12=opt_cbc.solve(model)
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print(data_csv['Drone Model'][i],' = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  120.0

Decision Variables
MANAS  =  10.0
JAY  =  0.0
VEER  =  0.0
SPASHT  =  0.0
DHAKSH  =  0.0

Constraints
constraints : Size=12
    Key : Lower : Body : Upper
     11 :  None : 60.0 :  61.0
     12 :  None : 42.5 :  91.0


The optimal value decreases as we go from pair 1 to 4 because from pair 1 to 4 the budget is decreasing.

